### PR TITLE
Make __getitem__ TINYer

### PR DESCRIPTION
**Pull request by geohotstan**
cleean refactooooooooooooors only (maybe)
still potential lines to be reduced in fancy indexing.
wip
wip
wip

### CODE REVIEW

**Comment by geohot** [comment]
Are all the exception cases tested?

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/tensor.py                        13              21
------------------------------------------------------------
lines added in the tinygrad folder: -8
```

**Comment by geohotstan** [comment]
> Are all the exception cases tested?

I didn't delete any of the exception cases but I did move them around a little. :s
I'll check and if they're not tested, I'll add tests.

**Comment by geohot** [comment]
Does GPT-4 have any good ideas?

https://chat.openai.com/share/30ab0b7e-d72c-444f-a228-620b132e9500

**Comment by geohotstan** [comment]
idk if collapsing the reshapes into 1 liners was the right choice for fancy indexing since we added 1 extra loop as well as we now have conditional checks within the loops (which could be mitigated with a `list + list` but then that's quite a long 1 liner...

and line 321 is totally unreadable

I added quite a lot of comments here... I'm not sure if it makes it more clear

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -265,78 +265,66 @@ def pad(self, arg: Tuple[Tuple[int, int], ...], value:float=0) -> Tensor:
...
-        new_idx = idx_.reshape(*[1]*dim[0], *idx_.shape, *[1]*(ret.ndim-dim[0]-idx_.ndim))
-        arange = Tensor.arange(ret.shape[d], dtype=dtypes.int32, requires_grad=False, device=self.device).reshape(*[1]*(d), ret.shape[d], *[1]*(ret.ndim-d-1))
-        ret = ((new_idx == arange) * ret).sum(d)
-      if dim[0] != 0 and dim != list(range(dim[0], dim[-1]+1)) and len(dim) != 1: # special permute case
-        order = list(range(ret.ndim))
-        order = order[dim[0]:dim[0]+idx[0].ndim] + order[:dim[0]] + order[dim[0]+idx[0].ndim:]
-        ret = ret.permute(order=order)
+      sum_dim = [d if n==0 else d+max_dim-n for n,d in enumerate(dim)]
+      arange = [Tensor.arange(ret.shape[d], dtype=dtypes.int32, requires_grad=False, device=self.device).reshape(*[1]*sd, ret.shape[d], *[1]*(((ret.ndim + max_dim) if n==0 else (ret.ndim + max_dim - n)) - sd - 1)) for n,(sd,d) in enumerate(zip(sum_dim, dim))]
+      idx = [i.reshape(*[1]*sum_dim[0], *((1, *[1]*(max_dim-i.ndim), *i.shape, *[1]*(ret.ndim - sd - 1)) if n == 0 else (*[1]*(max_dim-i.ndim), *i.shape, *[1]*(ret.ndim - sum_dim[0] - n)))) for n,(sd,i) in enumerate(zip(sum_dim, idx))]
```
Comment (wozeparrot) [inline tinygrad/tensor.py:321]: this is pretty unreadable, can you split this line up a bit?

Comment (geohotstan) [reply tinygrad/tensor.py]: Really happy with the new arange change. 
Current arange is readable and I separated idx 1 liner to 3 lines. Probably more readable

File: test/test_ops.py [hunk 2]
```diff
@@ -574,11 +574,16 @@ def test_slice_in_bounds_multidim(self):
     helper_test_op([(3,3,3)], lambda x: x[1:2, 1:2], lambda x: x[1:2, 1:2])
     helper_test_op([(3,3,3)], lambda x: x[1:2, 1:2, 0:-1], lambda x: x[1:2, 1:2, 0:-1])
 
-  def test_slice_with_none(self):
+  def test_slice_with_none_and_npnewaxis(self):
```
Comment (geohot) [inline test/test_ops.py:577]: Why is np.newaxis supported? This should be the same as None

Comment (geohotstan) [reply test/test_ops.py]: `  # - Indexing with np.newaxis or None on a given axis will add a new dimension of size one before that axis`
this line in the description of `_getitem_` mentioned npnewaxis and I didn't see it tested xD

But `np.newaxis` just evaluates to `None` when passed in as arg to `_getitem_`, so really they literally are the same thing.
Yeah, you're right, I'll also del the np.newaxis in the description. 
People should just use None instead.

**Comment by geohot** [comment]
This good to merge?

**Comment by geohotstan** [comment]
> This good to merge?

good on my end :D
